### ISSUE TITLE
Enhance connection logging and retry delay

### ISSUE THREAD

**Issue by hashi825** [NetApp/harvest#427]
**Is your feature request related to a problem? Please describe.**
Currently I experience occasional "target unreachable" errors on some collectors/pollers with no clear indication of visible network problems on the cluster or harvest side. These collectors exit standby on their next schedule but there's no clear error logging around this.

**Describe the solution you'd like**
Retry delay is hard coded to a max of 1024s, if possible have this configurable if specified.
Harvest doesn't seem to log the actual error it encounters but only sets the status as "connection error"
https://github.com/NetApp/harvest/blob/06ef5394d3c541be6e2d9d4eb9ad582bbbbc3948/cmd/poller/collector/collector.go#L305-L314

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by cgrinds** [comment]
thanks @hashi825 good suggestion. we'll improve this

**Comment by cgrinds** [comment]
hey @hashi825 here's a summary of the error handling in `collector.go`

# Logging

There are three kinds of errors that are treated special and then a catch-all for everything else.
The catch-all logs the err, but sounds like you're more interested in "target unreachable".

"Target unreachable" is one of the special cases, and as you note, there's a back-off since this error typically happens because of connection errors. `ERR_CONNECTION` is returned from `InitVars` (not relevant in this case) and zapi client `invoke` and `InvokeRaw`. The zapi client passes along the original error. 

Part of the reason ERR_CONNECTION is treated differently is because blind logging of that error when the cluster isn't reachable is chatty since each object is a separate collector (e.g. the ZapiPerf collector would log 19 lines, one for Processor, SystemNode, LUN, Volume, etc. The Zapi collector would log 8 lines per retry)

This code originally did more logging, but was reduced because of that chattiness.

What if we include the err in the log message at line 311, like so, would that be sufficient? 

```
me.Logger.Warn().
    Int("retryDelay", retryDelay).
    Err(err).
    Msg("Target unreachable, entering standby mode (retry in retryDelay s)")
```

Here's an example log message with that change (after dropping my network connection)

```
12:15PM WRN cmd/poller/collector/collector.go:317 > Target unreachable, entering standby mode (retry in retryDelay s) error="connection error => Post \"https://172.2.1.1:443/servlets/netapp.servlets.admin.XMLrequest_filer\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" Poller=venkat collector=ZapiPerf:FCVI retryDelay=1024
```

# Retry

The retry logic today does an exponential back off by increasing the retry by a factor of 4 (1s 4s 16s 64s 256s and finally 1024s) The retry delay won't increase beyond 1024s (17 minutes).

Are you wanting to lower or increase the maximum? Before we add another option I'd like to understand your usecase. Perhaps we should change the default? Are you interested in setting a different max, using a linear back-off, something different?

**Comment by hashi825** [comment]
I do remember seeing that kind of logging in an earlier build.

Basically in an earlier build of harvest I was experiencing these failures due to connection errors pretty frequently for no real reason
My current setup is two harvest instances polling the same clusters, they get scraped by different Prometheus instances. On the current build of harvest I largely have almost no connections errors anymore except for one cluster that will sporadically experience this from one of the harvest instances, this particular cluster is the most remote (50ms latency), I'm still trying to troubleshoot why this happens and maybe whether it's due to the polling happening at the same time but with the log verbosity it's hard to tell, maybe it could be enabled under debug?

As for the retry, a definable max would be good, as with my issue re-polling typically comes back fine. I couldn't find in the code where the retry delay gets reset after the collector recovers? If it doesn't do that then maybe fixing that would be better than allowing a configurable retry.

**Comment by vgratian** [comment]
The retry/delay is [here](https://github.com/NetApp/harvest/blob/7b0c5bd3ae8e7e2e837823da5b64c94297c27afc/cmd/poller/collector/collector.go#L305) in the code.

The current behavior of the collector:
- if connection errors, then it will retry after 4 seconds
- then, if cluster responds, it will go back to the normal interval
- if connection fails again, it will retry after 16 seconds (then 64s, then 256s, then it will keep retrying after 1024 seconds)

I think we could make this max value a parameter.

**Comment by vgratian** [comment]
E.g. we can add two parameters:
- `standby_increase_delay` (boolean) - weather or not to increase the delay
- `standby_increase_delay_max` (int) - how many times to increase the delay, when errors repeat

**Comment by cgrinds** [comment]
> I couldn't find in the code where the retry delay gets reset after the collector recovers?

Us either! Thanks for pointing that out - I fixed that in #436 and also added debug logs as your requested. Before we add more parameters for retry, let's see if #436 covers your case. I'm going to close this, please reopen if you end up needing retry parameters.
